---
bibliography: references.bib
biblio-style: apalike
highlight_bw: yes
output:
  bookdown::pdf_document2:
    toc: false
    includes:
      in_header: header.tex
    dev: "cairo_pdf"
    latex_engine: xelatex
    fig_caption: yes
geometry: margin=0.45in
link-citations: yes
---

**Supplemental File of**

\begingroup\Large
**ggtreeExtra: A universal package to visualize compact circular layers of phylogenetic tree**
\endgroup

**Shuangbin Xu, Zehan Dai, Pingfan Guo, Xiaocong Fu, Shanshan Liu, Lang Zhou, Wenli Tang, Tingze Feng, Meijun Chen, Li Zhan and Guangchuang Yu\* **

^\*^correspondence: guangchuangyu@gmail.com, gcyu1@smu.edu.cn

\renewcommand{\figurename}{Fig.}
\newcommand{\beginsupplement}{%
\setcounter{table}{0}
\renewcommand{\thetable}{S\arabic{table}}%
\setcounter{figure}{0}
\renewcommand{\thefigure}{S\arabic{figure}}%
}

\beginsupplement

```{r, echo=FALSE, message=FALSE, results='hide'}
require(kableExtra)
options(knitr.table.fromat = "latex")
knitr::opts_chunk$set(fig.pos= "!ht")
#knitr::opts_chunk$set(out.width="90%")
knitr::opts_chunk$set(fig.align="center")
usepackage_latex("float")
usepackage_latex("makecell")
usepackage_latex("booktabs")
```

```{r, echo=FALSE, message=FALSE}
library(ggtreeExtra)
library(ggtree)
library(treeio)
library(tidytree)
library(ggplot2)
library(ggnewscale)
library(patchwork)
```

# The purpose of development and overview

Integrating and visualizing associated data to the phylogenetic tree can help to find biological patterns and generate new hypotheses. The associated data type of phylogenetic tree can be roughly divided into continuous data and categorical data (discrete data). The continuous data sets represent measurements, they can be measured but not be counted, such as the height, weight, abundance of species, gene expression and the number of target genes etc. The categorical data sets represent characteristics, they can not be measured but they can be counted, such as endemic region information of virus, taxonomy information of species, type of target gene and sampling location information etc. Certainly, categorical data can also take on numerical values (for example, 1 for target gene A, 2 for target gene B). The associated data sets are also often multi-dimensional. Several tools have been developed to integrate and display associated data to phylogenetic tree. However, they still have some shortcomings, such as don't support annotation circular layout (tree and graphic alignment), provide few geometric layers, need predefined input etc, which make them not universal. Here, we developed *ggtreeExtra* to annotate multi-dimensional data to the outer of phylogenetic tree (Fig. \ref{fig:OverView}). It can link *ggtree*[@ggtree1] and geometric layers function defined in *ggplot2*[@ggplot2] or other ggplot2-based package. And it supports not only circular layout, but also other layouts defined in *ggtree*[@ggtree1]. The tree can be annotated by geometric function defined in ggtree[@ggtree1] (Fig. \ref{fig:OverView}). In addition, it was developed based on the grammar of graphics[@Wilkinson2012]. So user can easily map the variables (abundance of species, length of genome, sampling location) of associated data to aesthetic attributes (size, color, shape) of outer geometric objects (bar, point, box plot) of circular phylogenetic tree using *ggtreeExtra*. The details (such as legends and theme) of figure can be adjusted by corresponding *scale* function and theme function defined in ggplot2[@ggplot2] (Fig. \ref{fig:OverView}). Compared other tools, *ggtreeExtra* supports annotation of multiple layout of phylogenetic tree, it can integrate more geometric layers and don't need predefined input, which make it more universal (Tab. \ref{tab:compare}).

(ref:OverView) overview of ggtreeExtra. 

```{r, echo=FALSE, fig.cap="(ref:OverView)", fig.pos= "!h", out.width = '90%', OverView}
knitr::include_graphics("./overview.png", dpi=300)
```

```{r, echo=FALSE, message=FALSE, warning=FALSE, results='asis', compare}
require(kableExtra)

x <- "ggtreeExtra\tR,package\tcircular,inward circular,rectangular,slanted,ellipse,round rectangular,radial\tHeat map,scatter,simple bar,patter bar,stacked or dodged bar,box,pattern box,violin,dot intervals plot,density plot,pie,image plot\tYes\tYes\tprogramming\tadd,modify,delete\n
GraPhlAn\tPython,package\tcircular\tHeat map,scatter,simple bar\tNo\tYes\tcommand line,(configure file)\tadd\n
ETE3\tPython,package\trectangular\tHeat map,scatter,simple bar,stacked bar,box,pie,image plot\tNo\tYes\tprogramming,and interaction,(mouse click)\tadd\n
iTOL\tWeb,tool\tcircular,rectangular\tHeat map,scatter,simple bar,stacked bar,box,pie,image plot\tNo\tYes\tinteraction,(mouse click,configure file)\tadd\n
Microreact\tWeb,tool\tcircular,rectangular\tHeat map,scatter\tNo\tYes\tinteraction,(mouse click,and command line,configure file)\tadd\n
Evolview\tWeb,tool\tcircular,rectangular,slanted,round rectangular\tHeat map,scatter,simple bar,stacked bar,box\tNo\tYes\tinteraction,(mouse click,configure file)\tadd\n"

caption <- "Comparison list of ggtreeExtra and other tools"
header <- c("Tools", "Platform", "Supported layouts,for tree annotation[note]",
            "Annotation layers", "Supported,grammar,of graphic", "Combined freely[note]",
            "Methods of,figure out", "Layer,operations")

header <- gsub(" ", "\\ ", header) %>%
          linebreak(align="l", linebreaker = ",")

compare_tab <-
  read.table(text=x, header=F, sep="\t", quote="", check.names=F) %>%  
  dplyr::mutate(
         V3 = gsub(" ", "\\ ", V3),
         V4 = gsub(" ", "\\ ", V4),
         V7 = gsub(" ", "\\ ", V7)
  ) %>%
  dplyr::mutate(
         V3 = gsub(",", ",;", V3),
         V4 = gsub(",", ",;", V4),
         V8 = gsub(",", ",;", V8)
  )%>%
  dplyr::mutate(
         V2 = linebreak(V2, align = "l", linebreaker = ","),
         V3 = linebreak(V3, align = "l", linebreaker = ";"),
         V4 = linebreak(V4, align = "l", linebreaker = ";"),
         V7 = linebreak(V7, align = "l", linebreaker = ","),
         V8 = linebreak(V8, align = "l", linebreaker = ";")
  ) %>%
  knitr::kable(
               caption=caption,
               col.names=header,
               booktabs = T,
               position = "!h",
               escape=F
  ) %>%
  row_spec(0, bold=T) %>%
  row_spec(row=c(1,2,3,4,5), hline_after=T) %>%
  kable_styling(
                latex_options = c(
                                  "striped"#,
                #                  "scale_down"
                                  ),
                full_width = T,
                stripe_color = "gray!10",
                position="center"
  ) %>%
  add_footnote(c(
                 "tree annotation: tree and geometric alignment;",
                 "Combined freely: layers can be combined freely."
                 ),
               notation="number"
  ) %>% 
  landscape() %>%
  column_spec(c(1, 2), width = "4.5em", latex_valign="m") %>%
  column_spec(c(3, 4), width = "9.5em", latex_valign="m") %>%
  column_spec(7, width = "7em", latex_valign="m") %>%
  column_spec(c(5, 6, 8), width = "5em", latex_valign="m")

compare_tab
```

# Geometric layers that supported by *geom_fruit* of *ggtreeExtra*

*ggtreeExtra* is designed to link ggtree[@ggtree1] and some **geom** functions defined in *ggplot2*[@ggplot2] and other *ggplot2* extension packages (Fig. \ref{fig:OverView}). Here is the list of the geometric layer functions which work seamlessly with *geom_fruit* of *ggtreeExtra* (Tab. \ref{tab:geom}). Each geometric layer function has own unique geometric attributes (Tab. \ref{tab:geom}). User can choose appropriate geometric layer functions according the type of associated data. And the variables of associated data can be mapped to the attributes of corresponding geometric layer function. For example, when user want to view the distribution and uncertainty (continuous data) of associated data in different groups (such as the gene expression or species abundance in different samples), Box, violin or dot interval plot etc can be used to display them (Fig. \ref{fig:ggridges}). For the simple numeric data (length of genome, abundance of species), they can be visualized with simple (pattern) bar plot or grouped (pattern) bar plot (pattern bar plot is efficient to the visualization that don't want to use color) (Fig. \ref{fig:ggpattern}). Certainly, the aesthetics parameters of **geom** not only can be mapped by variables, but also can be used to adjust the attributes of geometric layers directly (such as *pattern_fill* in Fig. \ref{fig:ggpattern}). Since *ggtreeExtra* can work seamlessly with the *geom* functions, it can integrate more geometric layers compare other tools (Tab. \ref{tab:compare}). Until now, *ggtreeExtra* can integrate heat map, scatter plot, simple (grouped) (pattern) bar plot, simple (grouped) (pattern) box plot, violin, dot intervals plot, density plot, image plot, pie (Tab. \ref{tab:compare}). Even though other tools might provide the same graphic layers, *ggtreeExtra* is more flexible since it can inherit the features of the geometric functions, for example, *iTOL*, *Evolview* and *ggtreeExtra* all support displaying image plot on the phylogenetic tree, but *ggtreeExtra* is more flexible, it can integrate *ggplot2* object (Fig. \ref{fig:subplot}.A), and the attributes of subplot integrated can also be mapped to variable of associated data (Fig.\ref{fig:subplot}.B). As the *ggplot2*[@ggplot2] community keeps expanding and more *geom* functions will be implemented in either *ggplot2*[@ggplot2] or other extensions, *geom_fruit* will gain more power to present data in future.

```{r, echo=FALSE, results='asis', geom}

x <- "Package\tGeom layer\tVisual characteristic\tDescription\n
ggdist\tgeom_dots\talpha, color, fill, size, shape\tcreates dotplots that automatically determines a bin width that ensures the plot fits within the available space\n
ggdist\tgeom_dotsinterval\talpha, color, fill, size, shape\tcreates dots, intervals, and quantile dotplots\n
ggdist\tgeom_pointinterval\talpha, color, fill, size, shape\tcreates point and multiple uncertainty interval\n
ggdist\tgeom_slab\talpha, color, fill\tcreats slab geom\n
ggdist\tgeom_slabinterval\talpha, color, fill\tcreates slab, point and interval meta-geom\n
ggimage\tgeom_image\talpha, color, size\tvisualizes image files\n
ggimage\tgeom_phylopic\talpha, color, size\tqueries image files from phylopic database and visualizes them\n
ggpattern\tgeom_bar_pattern\tpattern_alpha, pattern_color, pattern_fill\tdraws bar charts with support for pattern fills\n
ggpattern\t \tpattern_angle, alpha, color, fill\t \n
ggpattern\tgeom_boxplot_pattern\tpattern_alpha, pattern_color, pattern_fill\tdraws box and whiskers plot with support for pattern fills\n
ggpattern\t \tpattern_angle, alpha, color, fill\t \n
ggpattern\tgeom_col_pattern\tpattern_alpha, pattern_color, pattern_fill\tdraws bar charts using ’stat_identity()’\n
ggpattern\t \tpattern_angle, alpha, color, fill\twith support for pattern fills\n
ggpattern\tgeom_tile_pattern\tpattern_alpha, pattern_color, pattern_fill\tdraws rectangle by using the center of the tile and its\n
ggpattern\t \tpattern_angle, alpha, color, fill\tsize with support for pattern fills\n
ggplot2\tgeom_bar\talpha, color, fill\tdraws bar charts\n
ggplot2\tgeom_boxplot\talpha, color, fill\tdraws box and whiskers plot\n
ggplot2\tgeom_col\talpha, color, fill\tdraws bar charts using 'stat_identity()'\n
ggplot2\tgeom_label\talpha, color, fill, size\tdraws a rectangle behind the text\n
ggplot2\tgeom_point\talpha, color, fill, shape, size\tcreats scatterplots\n
ggplot2\tgeom_raster\talpha, fill\ta high performance special case for all the tiles are the same size\n
ggplot2\tgeom_text\tcolor, size\tadds text to the plot\n
ggplot2\tgeom_tile\talpha, color, fill\tdraws rectangle by using the center of the tile and its size\n
ggpmisc\tgeom_plot\tvp.width,vp.height\tggplot objects as insets to the base ggplot, using syntax similar to that of 'geom_label'\n
ggpmisc\tgeom_table\tsize\tadds a textual table directly to the ggplot, using syntax similar to that of 'geom_label'\n
ggrepel\tgeom_text_repel\tcolor, size\tadds text to the plot. The text labels repel away from each other and away from the data points\n
ggrepel\tgeom_label_repel\talpha, color, fill, size\tdraws a rectangle underneath the text. The text labels repel away\n
ggrepel\t \t \tfrom each other and away from the data ponts\n
ggridges\tgeom_density_ridges\talpha,fill\tarranges multiple density plots in a staggered fashion\n
ggridges\tgeom_density_ridges2\talpha,fill\tarranges multiple density plots in a staggered fashion\n
ggridges\tgeom_ridgeline\talpha,color,fill\tplots the sum of the 'y' and 'height' aesthetics versus 'x', filling the area between 'y' and 'y + height' with a color\n
ggridges\tgeom_ridgeline_gradient\tcolor,fill\tworks just like 'geom_ridgeline' except that the 'fill' aesthetic can vary along the x axis\n
ggstance\tgeom_barh\talpha, color, fill\thorizontal version of 'geom_bar()'\n
ggstance\tgeom_boxploth\talpha, color, fill\thorizontal version of 'geom_boxplot()'\n
ggstance\tgeom_colh\talpha, color, fill\thorizontal version of 'geom_col()'\n
ggstar\tgeom_star\talpha, color, fill, size, starshape\tcreates scatterplots\n
ggsymbol\tgeom_symbol\talpha, color, fill, size, symbolshape\tcreates scatterplots\n
scatterpie\tgeom_scatterpie\talpha, color, fill\tcreates scatter pie plot\n
"

y <- read.table(text=x, sep="\t", header=T, quote="", check.names=F)
require(kableExtra)
caption = "List of geometric layers supported by 'geom\\textunderscore fruit()'"
knitr::kable(y, caption=caption, booktabs = T, position = "!h") %>%
  collapse_rows(columns = 1, latex_hline = "major", valign ="middle") %>%
  row_spec(0, bold=T) %>%
  kable_styling(
      latex_options = c(
                       "striped",
                       "scale_down"
                  ), 
      position="center")
```

(ref:ggridges) This example shows *ggtreeExtra* can work with *geom_density_ridges* of *ggridges*[@ggridges] and *geom_boxplot* of *ggplot2*[@ggplot2]. This dataset has also been visualized by *facet_plot*, but it can not work with circular layout tree[@ggtree2]. The associated data (melt_simple) was imported with *data* of *geom_fruit*, it has a column (*OTU*) contained tip labels. Then the tip labels column was assigned to *y*, the *val* is abundance of species (continuous data), which was mapped to the *x*, and *OTU* is the phylum information of species (categorical data), which was mapped to the *fill* (color).

```{r, fig.width=8, fig.height=6, error=TRUE, warning=FALSE, message=FALSE, fig.cap='(ref:ggridges)', dev="CairoPNG", dpi=300, ggridges}
library(ggtree)
library(ggplot2)
library(ggtreeExtra)
library(patchwork)
library(ggridges)
library(phyloseq)

set.seed(1024)
data("GlobalPatterns")
GP <- GlobalPatterns
GP <- prune_taxa(taxa_sums(GP) > 1000, GP)
sample_data(GP)$human <- get_variable(GP, "SampleType") %in%
  c("Feces", "Skin")

mergedGP <- merge_samples(GP, "SampleType")
mergedGP <- rarefy_even_depth(mergedGP, rngseed=1024)
mergedGP <- tax_glom(mergedGP,"Order")

melt_simple <- psmelt(mergedGP) %>%
  filter(Abundance < 120) %>%
  select(OTU, val=Abundance)

p <- ggtree(mergedGP, layout = "circular", size = 0.3) +
     geom_tippoint(aes(color = Phylum),
                   show.legend = FALSE,
                   size=0.6)

p1 <- p +
      geom_fruit(
          data = melt_simple,
          geom = geom_density_ridges,
          mapping = aes(
                      y = OTU,
                      x = val,
                      fill = Phylum,
                    ),
          offset = 0.12,
          pwidth = 0.4,
          lwd = .05,
          axis.params = list(axis       = "x",
                             text.size  = 1.2,
                             hjust      = 0,
                             vjust      = 1,
                             text.angle = -45
                        ),
          grid.params = list(),
          show.legend = FALSE
      ) +
      ggtitle("A") 

p2 <- p +
      geom_fruit(
          data = melt_simple,
          geom = geom_boxplot,
          mapping = aes(
                      y = OTU,
                      x = val,
                      fill = Phylum
                    ),
          offset = 0.12,
          pwidth = 0.4,
          size = 0.1,
          outlier.size = 0.4,
          outlier.stroke = 0.06,
          outlier.shape = 21,
          axis.params = list(
                          axis       = "x",
                          text.size  = 1.2,
                          hjust      = 0,
                          vjust      = 1,
                          text.angle = -45
                      ),
          grid.params = list(),
          show.legend = FALSE
      ) +
      ggtitle("B")

p1 + p2 
```

(ref:ggpattern) This example shows *ggtreeExtra* can work with *geom_bar_pattern* and *geom_boxplot_pattern* of *ggpattern*[@ggpattern]. This example also shows *ggtreeExtra* can work with multiple layouts of tree.

```{r, fig.width=7, fig.height=5, error=TRUE, warning=FALSE, message=FALSE, fig.cap='(ref:ggpattern)', dev="CairoPNG", dpi=300, ggpattern}
library(ggtree)
library(ggtreeExtra)
library(ggpattern)
library(ggplot2)
library(patchwork)

set.seed(1024)
tr <- rtree(20)

dat <- tibble::tribble(
            ~value, ~group,
            abs(rnorm(5, 10, sd = 3)), "A",
            abs(rnorm(3, 12, sd = 2)), "B",
            abs(rnorm(5, 9, sd = 3)),  "C",
            abs(rnorm(7, 6, sd = 2)),  "D"
       ) %>% tidyr::unnest(value)

dat$id <- tr$tip.label

dt <- tibble::tribble(
           ~value, ~class,
           abs(rnorm(40, 10, sd = 3)), "A",
           abs(rnorm(24, 12, sd = 2)), "B",
           abs(rnorm(40, 9, sd = 3)), "C",
           abs(rnorm(56, 6, sd = 2)), "D"
      ) %>% tidyr::unnest(value)

dt$id <- c(rep(tr$tip.label[1:5], 8), rep(tr$tip.label[6:8], 8),
           rep(tr$tip.label[9:13],8), rep(tr$tip.label[14:20], 8)
          )

p1 <- ggtree(tr, size=0.2, branch.length="none")
p2 <- ggtree(tr, size=0.2, layout="slanted", branch.length="none")
p3 <- ggtree(tr, size=0.2, layout="fan", open.angle=180, branch.length="none")
p4 <- ggtree(tr, size=0.2, layout="fan", open.angle=180, branch.length="none")

p1 <- p1 +
      geom_fruit(
          data=dat,
          geom=geom_bar_pattern,
          mapping=aes(y=id, x=value, pattern=group, pattern_angle=group),
          width=0.6, stat="identity",
          pwidth = 0.6, pattern_spacing =0.01,
          pattern_size = 0.1, pattern_density = 0.4,
          fill = "grey", pattern_fill="grey35",
          position=position_identityx(),
          axis.params=list(axis="x", text.size=1.5, text.angle=-45, hjust=0)
      ) + theme(legend.key.size = unit(0.3, 'cm'))

p2 <- p2 +
      geom_fruit(
          data=dat,
          geom=geom_bar_pattern,
          mapping=aes(y=id, x=value, pattern=group, pattern_fill=group),
          width=0.6, stat="identity",
          pwidth = 0.6, pattern_spacing =0.01,
          pattern_size = 0.1, pattern_density = 0.4,
          fill = "grey",
          position=position_identityx(),
          axis.params=list(axis="x",text.size=1.5, text.angle=-45, hjust=0)
      ) + theme(legend.key.size = unit(0.3, 'cm'))

p3 <- p3 +
      geom_fruit(
          data=dt,
          geom=geom_boxplot_pattern,
          mapping=aes(y=id, x=value, pattern=class, pattern_angle = class),
          size=0.1, outlier.shape=NA,
          pwidth=0.5, pattern_size = 0.1,
          pattern_density = 0.4, pattern_spacing =0.01,
          fill = "grey", pattern_fill="grey35",
          position=position_dodgex(),
          grid.params=list(),
          axis.params=list(axis="x", text.size=1.5, text.angle=-45, hjust=0)
      ) + 
      theme(legend.key.size = unit(0.35, 'cm'))

p4 <- p4 +
      geom_fruit(
          data=dt,
          geom=geom_boxplot_pattern,
          mapping=aes(y = id, x = value, pattern = class, pattern_fill = class),
          size = 0.1, outlier.shape = NA,
          pwidth = 0.5, pattern_size = 0.1,
          pattern_density = 0.4, pattern_spacing = 0.01,
          fill = "grey",
          position = position_dodgex(),
          grid.params = list(),
          axis.params = list(axis="x", text.size=1.5, text.angle=-45, hjust=0)
      ) + 
      theme(legend.key.size = unit(0.35, 'cm'))

(p1 + p2)/(p3 + p4)
```

(ref:subplot) This example show *ggtreeExtra* can integrate subplot to phylogenetic tree, and the subplot can be *ggplot* object (Fig.\ref{fig:subplot}.A), and the attributes of subplot can be mapped the variables of associated data when silhouette image file was used(Fig.\ref{fig:subplot}.B).

```{r, fig.width=7, fig.height=7, error=TRUE, warning=FALSE, message=FALSE, fig.cap='(ref:subplot)', dev="CairoPNG", dpi=300, subplot}
library(ggtreeExtra)
library(ggtree)
library(ggpmisc)
library(ggplot2)
library(ggimage)
library(patchwork)

set.seed(1024)
tr <- rtree(10)

dat1 <- data.frame(value=c(abs(rnorm(10, 3)),abs(rnorm(10, 5))), group=c(rep("A",10),rep("B",10)))
dat2 <- data.frame(value=c(abs(rnorm(15, 6)), abs(rnorm(15, 3))), group=c(rep("A",15),rep("B",15)))

subp1 <- ggplot(dat1) +
         geom_boxplot(aes(x=group, y=value, fill=group), 
                      size=0.1, outlier.size = 0.1, show.legend=F) +
         theme_bw() +
         theme(legend.key.size = unit(1, 'mm'),
               plot.background = element_rect(fill=NA, color=NA),
               panel.border = element_rect(size=0.1),
               legend.title = element_text(size=3.6),
               legend.text = element_text(size=3.2),
               axis.ticks = element_line(size=0.1),
               axis.text = element_text(size=2.8),
               axis.title = element_text(size=3))

subp2 <- ggplot(dat2) +
         geom_boxplot(aes(x=group, y=value, fill=group), 
         size=0.1, outlier.size = 0.1, show.legend=F) +
         theme_bw() +
         theme(legend.key.size = unit(1, 'mm'),
               plot.background = element_rect(fill=NA, color=NA),
               panel.border = element_rect(size=0.1),
               legend.title = element_text(size=3.6),
               legend.text = element_text(size=3.2),
               axis.ticks = element_line(size=0.1),
               axis.text = element_text(size=2.8),
               axis.title = element_text(size=3))

dt <- tibble::tibble(id=c("t1", "t2"), plot=list(subp1, subp2))

p1 <- ggtree(tr,
             layout="fan",
             open.angle=180
             ) +
      geom_tiplab(align=T, size=1.6) 

p2 <- p1 +
      geom_fruit(
                 data=dt,
                 geom=geom_plot,
                 mapping=aes(
                             y=id, 
                             label=plot
                         ),
                 offset=0.4,
                 position=position_identityx(),
                 vjust=0.6,
                 hjust=0.6,
                 vp.width=0.22,
                 vp.height=0.22
      ) +
      ggtitle("A")
 
newick <- "((Pongo_abelii,(Gorilla_gorilla_gorilla,(Pan_paniscus,Pan_troglodytes)
             Pan,Homo_sapiens)Homininae)Hominidae,Nomascus_leucogenys)Hominoidea;"

tree <- read.tree(text=newick)

p3 <- ggtree(tree,
             layout="fan",
             open.angle=180) +
      geom_tiplab(aes(label=label), offset = .2, size=1.6)

d <- ggimage::phylopic_uid(tree$tip.label)
d$body_mass <- c(52, 114, 47, 45, 58, 6)
d$ids <- tree$tip.label

p4 <- p3 +
      geom_fruit(
          data = d,
          geom = geom_phylopic,
          mapping = aes(y=ids, image=uid, colour=body_mass),
          position = position_identityx(),
          offset = 2,
          size = 0.04
      ) +
      ggtitle("B") +
      xlim(NA, 12) +
      scale_color_viridis_c(guide=guide_colorbar(barheight=2, barwidth=0.4)) +
      theme(legend.text=element_text(size=6),
            legend.title=element_text(size=7))

p2 + p4
```

# Examples of mapping and visualizing associated data on circular layout tree.

*ggtreeExtra* can integrate many geometric layers by linking *geom* function defined in *ggplot2*[@ggplot2] or ggplot2-extension packages (Tab. \ref{tab:geom} and \ref{tab:compare}). This is the basis for *ggtreeExtra* to be used for the annotation of phylogenetic tree. We presented two examples (Fig. \ref{fig:ggridges} and \ref{fig:ggpattern}) to show how to use *ggtreeExtra* to map simple associated data to phylogenetic tree by linking *geom* funtions. For the multiple associated data sets, circular layout is an efficient way to visualize multi-dimensional data sets, since it can reduce space and make the graph more compact. *ggtreeExtra* support annotation of multiple layouts of tree (Fig. \ref{fig:ggridges} and \ref{fig:ggpattern} and Tab. \ref{tab:compare}), including circular layout. Furthermore, the associated data can be imported with the *data* parameter of *geom_fruit* (Fig. \ref{fig:ggridges} and \ref{fig:ggpattern}), it can also be integrated to tree data (ggtree graphic object) with %<+% of *ggtree*[@ggtree2], before passing it to *geom_fruit* (Fig. \ref{fig:OverView}), or the tree data from output of the *treeio*, which can parse different file formats as well as the outputs of commonly used software[@treeio], can also be visualized by *ggtreeExtra*. Therefore, *ggtreeExtra* supports evolutionary statistics inferred by commonly used software, such as *BEAST*[@BEAST], *RAxML*[@RAxML], *HyPhy*[@HyPhy], *PAML*[@PAML], *ASTRAL*[@ASTRAL], *pplacer*[@pplacer], *RevBayes*[@RevBayes], PHYLODOG[@PHYLODOG] and EPA[@EPA] to be visualized by *ggtreeExtra*. Then the variables (abundance of species, length of genome, sampling location) of associated data can be mapped to the attributes of outer geometric objects (bar, point, boxplot) of circular phylogenetic tree (Fig. \ref{fig:OverView} and \ref{fig:ggridges}), since ggtreeExtra was developed based on grammar of graphic[@Wilkinson2012]. Here, we present several examples to elucidate how to map and display the associated data on the outer rings of circular phylogenetic trees using *ggtreeExtra*. More examples can be found on the *chapter10* of online book^[<http://yulab-smu.top/treedata-book>].

## Data supported by *ggtreeExtra*

*ggtreeExtra* is designed to work with *ggtree*[@ggtree1] and *treeio*[@treeio], it supports visualizing the tree data parsed by *treeio*[@treeio], any data frame that contains a column of tip labels can also be imported with *data* parameter of *geom_fruit* (Fig. \ref{fig:ggridges}). Or the data frame that contains a column of tip labels can also be integrated into *ggtree* graphic object using *%<+%* operator[@ggtree2], before passing it to *geom_fruit*. So *ggtreeExtra* can be used to annotate evolutionary statistics inferred by commonly used software, and It is possible to display the evolutionary statistics with phylogeny-associated data (such as traits, metadata) using *ggtreeExtra*. These features are not available in other tools. We reproduce Figure 3.3 of [@smith2019] to show how to integrate the associated data to tree data using *%<+%* and visualize it. And the source data is available at this repository^[<https://github.com/TheWrightonLab/Methanotroph_rpS3Analyses_SmithWrighton2018>]. The associated data sets contain the information of Ecosystem type, sequencing type and sample treatment method (all categorical data). The first column of external data is tip labels (the element of the column must be unique). The tree was built using *RAxML*[@RAxML], it is parsed to tree data using *read.raxml* of *treeio*[@treeio], contained bootstrap information. The tree data was visualized with *ggtree*[@ggtree1]. Then the associated data sets was imported *ggtree* graphic object using %<+%, *geom_fruit* can extract the tree data integrated automatically, and the attributes of data can be mapped using related attributes of geometry layers. The *y* of *aes* can be ignored. Here we used heat map to display the associated data sets (categorical data). The external ring heat maps represent the different types in corresponding categories (the type of Ecosystem was mapped to the color of innermost ring heat map, the type of sequencing was mapped to the color of middle ring heat map, the type of sample treatment was mapped to the color of outermost ring heat map) (Fig. \ref{fig:Metha}). In addition, compared with other tools, no restriction of data types should be visualized in *geom_fruit* of *ggtreeExtra*. It depends on the data types of various geometric functions (Tab. \ref{tab:geom}). For example, The multiple density plots is not supported by other tools, but *ggtreeExtra* can support it by linking *geom_density_ridges* of *ggridges*[@ggridges] (the data types of *geom_density_ridges* should be provided) (Fig. \ref{fig:ggridges}.A). The *ggplot* graphic object (one type of image plot) is also not supported by other tools, but *ggtreeExtra* can support it by linking *geom_plot* of *ggpmisc*[@ggpmisc], only we provides the data types of *geom_plot* (Fig. \ref{fig:subplot}.A). As the *geom* of *ggplot2* [@ggplot2] or other extensions will be updated and developed, so *ggtreeExtra* will be also more to present datasets in the future.

(ref:Metha) Phylogeny of methanotroph ribosomal protein S39 (rpS3) genes from Figure 3.3 [@smith2019]. The external ring of circular tree is built with the associated data sets integrated to tree data using *%<+%*.

```{r, fig.width=7, fig.height=7, error=TRUE, warning=FALSE, message=FALSE, fig.cap='(ref:Metha)', dev="CairoPNG", dpi=300, Metha}
library(ggtreeExtra)
library(ggtree)
library(treeio)
library(ggplot2)
library(ggnewscale)
tree <- read.raxml("../data/Methanotroph/Methanotroph_rpS3_Modified_Alignment_RAxML")
# Root the tree to the archaea sequences
tree@phylo <- root(tree@phylo, node=1402, edgelabel=TRUE)
df <- read.csv("../data/Methanotroph/metadata.csv")
# reset the levels of columns to reproduce the order of original figure.
df$Specific.Ecosystem <- factor(df$Specific.Ecosystem,
                                levels=c("Agriculture", "Alkaline/Hypersaline",
                                         "Contaminated/Wastewater", "Endosymbiont",
                                         "Freshwater", "Forest", "Geothermal",
                                         "Marine", "Natural Seep", "Peat",
                                         "Permafrost", "Wetland", "Unknown"))
df$MetaType <- factor(df$MetaType,
                     levels=c("Metatranscriptome", "Metagenome",
                              "Single-amplified genome", "Fosmid", NA))
df$Treatment <- factor(df$Treatment, levels=c("Native", "Enrichment","Isolate", "Unknown"))

p <- ggtree(tree, layout="fan", open.angle=30)
p <- rotate_tree(p, 90)
p1 <- p + geom_treescale(x=0.2, y=727*6/11, width=1, offset=20) +
      geom_nodepoint(aes(subset=as.numeric(bootstrap)>=70),
                     shape=21, fill="grey", size=1)
print(as.treedata(p1))
# we can use %<+% to integrate the external datasets to tree structure.
# and the y can not be specified.
p2 <- p1 %<+% df +
      geom_fruit(
          geom=geom_tile,
          mapping=aes(fill=Specific.Ecosystem),
          offset=0.13,
          width=0.35,
          axis.params=list(axis="x", text="Ecosystem", text.angle=0,
                          hjust=0, text.size=3, family="Times", fontface="bold")
     ) +
     scale_fill_manual(
         values=c("green3","turquoise", "maroon", "orchid",
          "deepskyblue", "forestgreen","salmon", "cadetblue3",
          "slategray4",  "yellowgreen", "gray90", "chocolate2",
          "yellow"),
         guide=guide_legend(title="Ecosystem", keywidth=0.5, keyheight=0.5, order=4),
         na.translate=FALSE
     ) +
     new_scale_fill() +
     geom_fruit(
         geom=geom_tile,
         mapping=aes(fill=MetaType),
         offset=0.13,
         width=0.35,
         axis.params=list(axis="x", text="Sequencing Type", text.angle=0,
                         hjust=0, text.size=3, family="Times", fontface="bold")
     ) +
     scale_fill_manual(
         values=c("red", "black", "dodgerblue", "gray50"),
         guide=guide_legend(title="Sequencing Type", keywidth=0.5, keyheight=0.5, order=3),
         na.translate=FALSE
     ) +
     new_scale_fill() +
     geom_fruit(
         geom=geom_tile,
         mapping=aes(fill=Treatment),
         offset=0.13,
         width=0.35,
         axis.params=list(
                         axis="x",
                         text="Sample Treatment",
                         text.angle=0,
                         hjust=0,
                         text.size=3,
                         family="Times",
                         fontface="bold"
                     )
     ) +
     scale_fill_manual(
         values=c("red","gray50", "black", "yellow"),
         guide=guide_legend(
                   title="Sample Treatment",
                   keywidth=0.5,
                   keyheight=0.5,
                   order=2),
         na.translate=FALSE
     ) +
     theme(
           legend.background=element_rect(fill=NA), # the background of legend.
           legend.title=element_text(size=9, family="Times", face="bold"),
           legend.text=element_text(size=7, family="Times"), # the text size of legend.
           legend.spacing.y = unit(0.02, "cm"),
           legend.margin=margin(0.1, 0.9, 0.1,-0.9, unit="cm"), # t, r, b, l, cm
           legend.box.margin=margin(0.1, 0.9, 0.1, -0.9, unit="cm"),
           plot.margin = unit(c(-1.2, -1.2, -1.2, 0.1),"cm")
     )

# optional
cladeda <- data.frame(nodeid = c(793, 791, 1384, 1394, 1440, 1405), 
                      label = c("Gammaproteobacteria", "Alphaproteobacteria",
                              "Ca.Methylomirabilis", "Methylacidiphilae",
                              "ANME-1", "ANME-2"), 
                      horizontal = c(FALSE, FALSE, TRUE, TRUE, TRUE, TRUE),
                      hjust = c(0.5, 0.5, 0, 0, 0, 0))

p3 <- p2 + 
      geom_cladelab(
          data = cladeda,
          mapping = aes(node=nodeid, label=label, horizontal=horizontal, hjust=hjust),
          angle = "auto",
          offset = 1.4, 
          align = T, 
          fontsize = 2, 
          barsize = 1,
          family = "Times",
          fontface="bold",
          offset.text = 0.1
      )
p3
```

## Displaying multiple associated data to circular phylogenetic tree using grammar of graphic

*ggtreeExtra* is developed based on grammar of graphic[@Wilkinson2012]. The variables of associated data sets can be mapped to the attributes of geometric layers, the graphic can be displayed with provided aesthetic mapping and non-variable setting. And users can specify their own set of mappings from levels or values in the data to aesthetic values using the corresponding *scale* function defined in *ggplot2* or *ggplot2-based* packages (Fig. \ref{fig:OverView}). They does not need providing configure file which mixes associated data and the profiler of graphic (it is needed by many other tools (Tab. \ref{tab:compare})), so no restriction of data types should be used in *geom_fruit* of *ggtreeExtra* (see 3.1 section). Here, we reproduce Fig.2 of [@morgan2013HMP] to show how to display multiple associated data to circular phylogenetic tree using *ggtreeExtra*. The data sets are provided by GraPhlAn [@GraPhlAn], but they were mixed with the profiler of graphic for GraPhlAn. We has extracted and saved them as the data frame tables for corresponding geometric layer functions. These data sets contain the relative abundance of bacteria (continuous data) at different body sites (categorical data). The associated data sets were imported with *data* parameter of *geom_fruit* and displayed with the corresponding geometric function. We used heat map to display the abundance of tip species by linking *geom_tile* of *ggplot2*[@ggplot2] (the abundance of species was mapped to the transparency of heat map using *alpha* aesthetic of *geom_tile*, then *alpha* was specified the mapping from values in the data using *scale_alpha_continuous* defined in *ggplot2*. The different body sites was mapped to the color of heat map using *fill* aesthetic of *geom_tile*, then the *fill* was specified the mapping from levels in the data using *scale_fill_manual* defined in *ggplot2*) (Fig. \ref{fig:HMP}.C), and the most abundance of species (continuous data) at specific site was visualized with bar plot by linking *geom_col* of *ggplot2* (the abundance of species was mapped to the length of bar plot using *x* aesthetic of *geom_col*, the different body sites was mapped to the color of bar plot using *fill* aesthetic of *geom_col*.) (Fig. \ref{fig:HMP}.D). The outer graphic layers can be aligned automatically according to the tip labels of phylogenetic tree (tip labels were mapped to *y* value). The tree annotated by *geom_fruit* can also be annotated beforehand with other geometric layers provided by *ggtree*[@ggtree1], *ggplot2* or ggplot2-based packages, such as *geom_star* of *ggstar*[@ggstar] (Fig. \ref{fig:HMP}.A), *geom_hilight* and *geom_cladelab* of *ggtree* (Fig. \ref{fig:HMP}.B). The layers and outer graphic layers can be added step-by-step with **+** symbol (Fig. \ref{fig:HMP}). More step-by-step instructions are also available at the vignette^[<http://bioconductor.org/packages/release/bioc/vignettes/ggtreeExtra/inst/doc/ggtreeExtra.html>].

(ref:HMP) The abundance of microbes at different sites of human. The shape of tip labels indicated the commensal microbes or potential pathogens. The transparency of heat map indicates the abundance of microbes, and colors of heat map indicate the different sites of human. The bar plot indicates the relative abundance at body site of the most abundance. Fig.\ref{fig:HMP}.A contains the tree layer and tip point layer; Fig.\ref{fig:HMP}.B is from Fig.\ref{fig:HMP}.A added high light layer and clade label layer; Fig.\ref{fig:HMP}.C is from Fig.\ref{fig:HMP}.B added a heatmap layer; Fig.\ref{fig:HMP}.D is from Fig.\ref{fig:HMP}.C added a bar chart layer. This example shows the abilities of annotate phylogenetic tree through grammar of graphic using ggtree and ggtreeExtra.

```{r, fig.width=8, fig.height=8, error=TRUE, warning=FALSE, message=FALSE, fig.cap='(ref:HMP)', dev="CairoPNG", dpi=300, HMP}
library(ggtreeExtra)
library(ggtree)
library(treeio)
library(tidytree)
library(ggstar)
library(ggplot2)
library(ggnewscale)
library(patchwork)

tree <- read.tree("../data/HMP_tree/hmptree.nwk")
# the abundance and types of microbes
dat1 <- read.csv("../data/HMP_tree/tippoint_attr.csv")
# the abundance of microbes at different body sites.
dat2 <- read.csv("../data/HMP_tree/ringheatmap_attr.csv")
# the abundance of microbes at the body sites of greatest prevalence.
dat3 <- read.csv("../data/HMP_tree/barplot_attr.csv")

# adjust the order
dat2$Sites <- factor(dat2$Sites, levels=c("Stool (prevalence)", "Cheek (prevalence)",
                                          "Plaque (prevalence)","Tongue (prevalence)",
                                          "Nose (prevalence)", "Vagina (prevalence)",
                                          "Skin (prevalence)"))
dat3$Sites <- factor(dat3$Sites, levels=c("Stool (prevalence)", "Cheek (prevalence)",
                                          "Plaque (prevalence)", "Tongue (prevalence)",
                                          "Nose (prevalence)", "Vagina (prevalence)",
                                          "Skin (prevalence)"))
# extract the clade label information. Because some nodes of tree are annotated to genera,
# which can be displayed with high light using ggtree.
nodeids <- nodeid(tree, tree$node.label[nchar(tree$node.label)>4])
nodedf <- data.frame(node=nodeids)
nodelab <- gsub("[\\.0-9]", "", tree$node.label[nchar(tree$node.label)>4])
# The layers of clade and hightlight (optional)
poslist <- c(1.6, 1.4, 1.6, 0.8, 0.1, 0.25, 1.6, 1.6, 1.2, 0.4,
             1.2, 1.8, 0.3, 0.8, 0.4, 0.3, 0.4, 0.4, 0.4, 0.6,
             0.3, 0.4, 0.3)
labdf <- data.frame(node=nodeids, label=nodelab, pos=poslist)
# The circular layout tree.
p <- ggtree(tree, layout="fan", size=0.15, open.angle=5) 
p <- p %<+% dat1 +
     geom_star(
          mapping=aes(fill=Phylum, starshape=Type, size=Size),
          starstroke=0.05
     ) +
     scale_fill_manual(
          values=c("#FFC125","#87CEFA","#7B68EE","#808080","#800080",
                   "#9ACD32","#D15FEE","#FFC0CB","#EE6A50","#8DEEEE",
                   "#006400","#800000","#B0171F","#191970"),
          guide=guide_legend(keywidth = 0.5, keyheight = 0.5, order=1,
          override.aes=list(starshape=15)),
          na.translate=FALSE
     ) +
     scale_starshape_manual(
          values=c(15, 1),
          guide=guide_legend(keywidth = 0.5, keyheight = 0.5, order=2),
          na.translate=FALSE
     ) +
     scale_size_continuous(
          range = c(0.5, 1.5),
          guide = guide_legend(keywidth = 0.5, keyheight = 0.5, order=3,
          override.aes=list(starshape=15))
     ) +
     new_scale_fill() +
     ggtitle("A") +
     theme(legend.position="none")

# optional for high light and clade labels     

p1 <- p + 
      geom_hilight(
         data=nodedf,
         mapping=aes(node=node),
         extendto=6.8,
         alpha=0.3,
         fill="grey",
         color="grey50",
         size=0.05
      ) +
      geom_cladelab(
         data=labdf,
         mapping=aes(node=node, label=label, offset.text=pos),
         barsize=NA,
         fontsize=0.7,
         angle="auto",
         hjust=0.5,
         horizontal=FALSE,
         fontface="italic"
      ) +
      ggtitle("B")

p2 <- p1 +
      geom_fruit(
          data=dat2,
          geom=geom_tile,
          mapping=aes(y=ID, x=Sites, alpha=Abundance, fill=Sites),
          color = "grey50",
          offset = 0.04,
          size = 0.02
      )+
      scale_alpha_continuous(
          range=c(0, 1),
          guide=guide_legend(keywidth = 0.3, keyheight = 0.3, order=5)
      ) +
      scale_fill_manual(
          values=c("#0000FF","#FFA500","#FF0000","#800000",
                   "#006400","#800080","#696969"),
          guide=guide_legend(keywidth = 0.3, keyheight = 0.3, order=4)
      ) +
      ggtitle("C") +
      theme(legend.position="none")

p3 <- p2 +
      geom_fruit(
          data=dat3,
          geom=geom_col,
          mapping=aes(y=ID, x=HigherAbundance, fill=Sites),
          pwidth=0.38,
          orientation="y",
          position=position_stackx(),
      )+
      geom_treescale(fontsize=1.2, linesize=0.3, x=4.9, y=0.1) +
      ggtitle("D") +
      theme(legend.position="none")

p4 <- (p + p1 + plot_layout(width=c(3.4,4)))/ (p2 + p3 + plot_layout(width=c(3.4,4))) +
      plot_layout(heights=c(3,4))

p4
```

## Annotating phylogenetic tree fully with multiple associated data

Since *ggtreeExtra* supports annotation of multiple layouts tree (tree and geometric layers alignment) (Tab. \ref{tab:compare} and Fig. \ref{fig:ggridges} and \ref{fig:ggpattern}). It can also link *ggtree*[@ggtree1] and geometric layer functions defined in *ggplot2*[@ggplot2] or ggplot2-based packages based on grammar of graphic (Tab. \ref{tab:geom} in **2 section** and Fig. \ref{fig:HMP} in **3.2 section**). The annotation layers on the outer of phylogenetic tree can be combined freely. So it can be easily used to visualize complex and high-dimensional associated data on the outer of phylogenetic tree including the tree containing thousands of tips. Here, we reproduce Fig 2 of [@GraPhlAn]. The phylogenetic tree also was built using a part of 3737 microbes [@PhyloPhlAn]. The associated data sets have also been extracted from configure file and saved as the data frame types for corresponding geometric layer functions. They contains present or not (discrete data) and type (discrete data) of target gene, the type (discrete data) and capability (continuous data) of fatty acid metabolism, and the length of microbes genome. The present or not of different target gene was visualized with first inner heat map (the type of target gene was mapped to color of heat map, and the subtype of target gene was mapped to the x value(take on numerical values)). The type and capability of fatty acid metabolism was also visualized with middle heat map (the type was mapped to color of heat map, the capability of fatty acid metabolism was mapped to the transparency of heat map). The length of genome was displayed with bar chart (the length of genome was mapped to length of bar, the phylum information of genome was mapped to the color of bar). (Fig. \ref{fig:KEGG}). Notably, the same colors of Proteobacteria and Spirochaetes were found in the original configure file[@GraPhlAn]. We tried to separate the tips from Proteobacteria and Spirochaetes. Unfortunately, the phylum information of tips was replaced color code, so color of Proteobacteria and Spirochaetes is also identical in Fig. \ref{fig:KEGG}. This show that providing configure files mixed associated data and profiler of graphic layer is tedious and error-prone since the profilerd of graphic layer (such like color or size) should be consistent for the same information. 

(ref:KEGG) The phylogenetic tree built using in [@PhyloPhlAn] using 963 microbial genomes (a part of 3737 microbes [@PhyloPhlAn]). The first and second ring heat map were built with discrete data, it represents the presence or absence of each module, the other heat map rings were created with continuous data, the transparency represents the capability of fatty acid metabolism, the different colors represent the types of fatty acid metabolism. The length of bar represents the genome length of corresponding microbes.

```{r, fig.width=7, fig.height=7, error=TRUE, warning=FALSE, message=FALSE, fig.cap="(ref:KEGG)", dev="CairoPNG", dpi=300, KEGG}
library(ggtreeExtra)
library(ggtree)
library(treeio)
library(tidytree)
library(ggstar)
library(ggplot2)
library(ggnewscale)

tree <- read.tree("../data/kegg/kegg.nwk")
# The attributes of tip point
dt1 <- read.csv("../data/kegg/tippoint_attr.csv")
# The attributes of first ring
dt2 <- read.csv("../data/kegg/firstring_attr.csv")
# The attributes of second ring
dt3 <- read.csv("../data/kegg/secondring_attr.csv")
# The attrihutes of bar plot
dt4 <- read.csv("../data/kegg/barplot_attr.csv")

#reorder the Phyla column
dt1$Phyla <- factor(dt1$Phyla, levels=c("Actinobacteria","Aquificae","Bacteroidetes",
                                        "Chlamydiae","Chlorobi","Chloroflexi","Crenarchaeota",
                                        "Cyanobacteria","Euryarchaeota","Firmicutes","Proteobacteria",
                                        "Spirochaetes","Tenericutes","Thermi","Thermotogae","Other"))

# reorder the Type2 column
dt3$Type2 <- factor(dt3$Type2, levels=c("FA synth init", "FA synth elong",
                                        "acyl-CoA synth", "beta-Oxidation",
                                        "Ketone biosynth"))

dt4$Phyla <- factor(dt4$Phyla, levels=c("Actinobacteria","Aquificae","Bacteroidetes",
                                        "Chlamydiae","Chlorobi","Chloroflexi","Crenarchaeota",
                                        "Cyanobacteria","Euryarchaeota","Firmicutes","Proteobacteria",
                                        "Spirochaetes","Tenericutes","Thermi","Thermotogae","Other"))
# extract node label for the clade layers
nodelab <- tree$node.label[nchar(tree$node.label)>0]
nodeids <- nodeid(tree, nodelab)

# the position of clade label
textex <- c(1.0, 0.4, 0.2, 1.4, 1.4, 0.4, 1.4, 1.4, 0.4, 0.4,
             0.8, 1, 0.6, 0.6, 0.4, 0.3, 0, 0.4, 0.1, 0.25,
             0.2, 0.3, 0.8, 0.8, 0.8, 0.6, 2.4)
# optional for clade layers
cladelabels <- mapply(function(x, y, z){geom_cladelabel(node=x, label=y, barsize=NA, extend=0.3,
                                              offset.text=z, fontsize=1.2, angle="auto",
                                              hjust=0.5, horizontal=FALSE, fontface="italic")},
                                     nodeids, nodelab, textex, SIMPLIFY=FALSE)

# high light layers
fills <- c("#808080", "#808080", "#808080", "#808080", "#808080",
           "#191970", "#87CEFA", "#FFC125", "#B0171F", "#B0171F",
           "#B0171F", "#B0171F", "#B0171F", "#B0171F", "#B0171F",
           "#B0171F", "#B0171F", "#B0171F", "#B0171F", "#B0171F",
           "#B0171F", "#B0171F", "#9ACD32", "#9ACD32", "#9ACD32",
           "#006400", "#800000")
# optional for hight light
highlights <- mapply(function(x, y){geom_hilight(node=x, extendto=5.8, alpha=0.3,
                                                 fill=y, color=y, size=0.05)},
                      nodeids, fills, SIMPLIFY=FALSE)

# to reproduce the original figures, we use the same colors.
# uses can custom set it.
colors <- c("#9ACD32", "#EE6A50", "#87CEFA", "#FFC125", "#D15FEE", "#8DEEEE", "#800000",
            "#006400", "#800080", "#808080", "#B0171F", "#B0171F", "#191970", "#7B68EE",
            "#00CD00", "Black")

p1 <- ggtree(tree, layout="circular", size=0.1) + highlights

p2 <- p1 +
     geom_fruit(
         data=dt1,
         geom=geom_point,
         mapping=aes(y=ID, fill=Phyla),
         shape=21,
         size=1.2,
         stroke=0.05,
         position="identity",
         show.legend=FALSE
    )+
    scale_fill_manual(values=colors) +
    cladelabels +
    new_scale_fill() 

p3 <- p2 +  
      geom_fruit(
        data=dt2,
        geom=geom_tile,
        mapping=aes(y=ID, x=ring, fill=Type1),
        offset=-0.02,
        pwidth=0.14,
        addbrink=TRUE
    ) +
    scale_fill_manual(
        name="ATP synthesis",
        values=c("#339933", "#dfac03"),
        guide=guide_legend(keywidth=0.5, keyheight=0.5, order=1)
    ) +
    new_scale_fill() 

p4 <- p3 + 
      geom_fruit(
          data=dt3,
          geom=geom_tile,
          mapping=aes(y=ID, alpha=Abundance, x=Type2, fill=Type2),
          offset=0.001,
          pwidth=0.18
      ) +
      scale_fill_manual(
          name="Fatty Acid metabolism",
          values=c("#b22222", "#005500", "#0000be", "#9f1f9f", "#793a07"),
          guide=guide_legend(keywidth=0.5, keyheight=0.5, order=2)
      ) +
      scale_alpha_continuous(range=c(0, 0.4),
                             guide=guide_legend(keywidth=0.5, keyheight=0.5,order=3)) +
      new_scale_fill()

p5 <- p4 +
      geom_fruit(data=dt4,
           geom=geom_bar,
           mapping=aes(y=ID, x=Length, fill=Phyla),
           stat="identity",
           orientation="y",
           pwidth=0.3,
           position=position_dodgex()) +
      scale_fill_manual(values=colors,
                        guide=guide_legend(keywidth=0.5, keyheight=0.5, order=4)) +
      geom_treescale(fontsize=1.2, linesize=0.3) +
      theme(legend.position=c(0.95, 0.5),
            legend.background=element_rect(fill=NA),
            legend.title=element_text(size=7),
            legend.text=element_text(size=6),
            legend.spacing.y = unit(0.02, "cm"))
p5
```

## Annotating associated data to the phylogenetic tree combined chord diagram.


# Summary


# References
